packages/SwingSet/src/kernel/kernel.js

```diff
@@ -503,23 +510,93 @@ export default function buildKernel(kernelEndowments) {
     });
 
     // the vatManager invokes setup() to build the userspace image
-    const manager = makeVatManager(
+    return makeVatManager(
       vatID,
       syscallManager,
       setup,
       helpers,
       kernelKeeper,
-      vatKeeper,
-    );
-    ephemeral.vats.set(
-      vatID,
-      harden({
-        manager,
-        enablePipelining: Boolean(options.enablePipelining),
-      }),
+      kernelKeeper.provideVatKeeper(vatID),
     );
   }
 
+  // This kernel (and the genesis vats) will have been evaluated in a context in
+  // which these canonical symbols were available (which is why the require
+  // statements here work). That was achieved by running makeRequire() in
+  // controller.js. We want to provide the same context when we evaluate dynamic
+  // vats.
+  function kernelRequire(nameArg) {
+    const name = `${nameArg}`;
+    const HARDEN = '@agoric/harden';
+    const NAT = '@agoric/nat';
+    const EVALUATE = '@agoric/evaluate';
+    switch (name) {
+      case HARDEN:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(HARDEN);
+      case NAT:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(NAT);
+      case EVALUATE:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(EVALUATE);
+      default:
+        throw Error(`require ${name} is not supported`);
+    }
+  }
+
+  // Enqueue a message to the adminVat giving it the new vat's root object
+  function notifyAdminVatOfNewVat(vatID) {
+    function serializeSlot(slot) {
+      return {
+        body: `["${vatID}",{"@qclass":"slot","index":0}]`,
+        slots: [slot],
+      };
+    }
+    const vatSlot = makeVatRootObjectSlot();
+    const kernelRootObjSlot = addExport(vatID, vatSlot);
+    const serializedArgs = serializeSlot(kernelRootObjSlot);
+    const vatAdminVatId = vatNameToID('vatAdmin');
+    queueToExport(vatAdminVatId, vatSlot, 'newVatCallback', serializedArgs);
+  }
+
+  // Create a new vat and return the vatID without waiting for the root object.
```

Comment:
"without waiting for the root object": `buildFn` is called (and returns), so the root object has been created.. what are we not waiting for?

Maybe say "without registering the root object in the c-list; that will be added when `notifyAdminVatOfNewVat` calls `addExport`" ?